File: PolliLib/model_retrieval.py
"""
Model Retrieval - List available text and image models
Implements the Model Retrieval section from the TODO list
"""

from .pollylib import PollinationsAPI
import json
from typing import List, Dict, Any


class ModelRetrieval(PollinationsAPI):
    """Class for retrieving available models from Pollinations.AI"""

    def list_text_models(self, normalized: bool = True) -> List[Dict[str, Any]]:
        """
        List all available text generation models.

        Args:
            normalized: Return normalized model schema with full details

        Returns:
            List of model information dictionaries
        """
        try:
            response = self.retry_request(
                "GET",
                f"{self.TEXT_API}/models"
            )

            models = response.json()

            if normalized:
                # Return normalized schema as per TODO requirements
                return self._normalize_text_models(models)

            return models

        except Exception as e:
            print(f"Error retrieving text models: {e}")
            return []

    def _normalize_text_models(self, models: Any) -> List[Dict[str, Any]]:
        """
        Normalize text model data into standard schema.

        Returns schema with:
        - name and description
        - max input tokens
        - reasoning capability flag
        - tier
        - community supported flag
        - input types array
        - output types array
        - tool use / function calling flag
        - aliases array
        - vision flag
        - audio flag
        - voices array
        - system messages supported flag
        - uncensored flag
        """
        normalized = []

        # Handle different response formats
        if isinstance(models, list):
            model_list = models
        elif isinstance(models, dict):
            model_list = models.get('models', [])
        else:
            return []

        for model in model_list:
            if isinstance(model, str):
                # Basic model name only
                normalized_model = {
                    "name": model,
                    "description": f"{model} text generation model",
                    "max_input_tokens": 128000,  # Default estimate
                    "reasoning_capable": "reasoning" in model.lower(),
                    "tier": "standard",
                    "community_supported": False,
                    "input_types": ["text"],
                    "output_types": ["text"],
                    "tool_use": "openai" in model.lower(),
                    "aliases": [],
                    "vision": "vision" in model.lower() or model in ["openai", "openai-large", "claude-hybridspace"],
                    "audio": "audio" in model.lower(),
                    "voices": ["alloy", "echo", "fable", "onyx", "nova", "shimmer"] if "audio" in model.lower() else [],
                    "system_messages_supported": True,
                    "uncensored": False
                }
            else:
                # Structured model data
                normalized_model = {
                    "name": model.get("name", "unknown"),
                    "description": model.get("description", ""),
                    "max_input_tokens": model.get("max_input_tokens", 128000),
                    "reasoning_capable": model.get("reasoning_capable", False),
                    "tier": model.get("tier", "standard"),
                    "community_supported": model.get("community_supported", False),
                    "input_types": model.get("input_types", ["text"]),
                    "output_types": model.get("output_types", ["text"]),
                    "tool_use": model.get("tool_use", False),
                    "aliases": model.get("aliases", []),
                    "vision": model.get("vision", False),
                    "audio": model.get("audio", False),
                    "voices": model.get("voices", []),
                    "system_messages_supported": model.get("system_messages_supported", True),
                    "uncensored": model.get("uncensored", False)
                }

            normalized.append(normalized_model)

        return normalized

    def list_image_models(self, normalized: bool = True) -> List[Dict[str, Any]]:
        """
        List all available image generation models.

        Args:
            normalized: Return normalized model schema with full details

        Returns:
            List of model information dictionaries including:
            - style tags
            - input/output limits
            - supported formats
        """
        try:
            response = self.retry_request(
                "GET",
                f"{self.IMAGE_API}/models"
            )

            models = response.json()

            if normalized:
                return self._normalize_image_models(models)

            return models

        except Exception as e:
            print(f"Error retrieving image models: {e}")
            return []

    def _normalize_image_models(self, models: Any) -> List[Dict[str, Any]]:
        """
        Normalize image model data into standard schema.

        Returns schema with:
        - name and description
        - style tags
        - input/output limits
        - supported formats
        """
        normalized = []

        # Handle different response formats
        if isinstance(models, list):
            model_list = models
        elif isinstance(models, dict):
            model_list = models.get('models', [])
        else:
            return []

        # Known model characteristics
        model_info = {
            "flux": {
                "description": "High-quality image generation model",
                "style_tags": ["photorealistic", "artistic", "detailed"],
                "max_width": 2048,
                "max_height": 2048,
                "supports_img2img": False
            },
            "turbo": {
                "description": "Fast image generation model",
                "style_tags": ["quick", "artistic"],
                "max_width": 1024,
                "max_height": 1024,
                "supports_img2img": False
            },
            "kontext": {
                "description": "Image-to-image transformation model",
                "style_tags": ["transformation", "editing"],
                "max_width": 2048,
                "max_height": 2048,
                "supports_img2img": True
            }
        }

        for model in model_list:
            if isinstance(model, str):
                model_name = model
                info = model_info.get(model_name, {})

                normalized_model = {
                    "name": model_name,
                    "description": info.get("description", f"{model_name} image model"),
                    "style_tags": info.get("style_tags", ["general"]),
                    "max_width": info.get("max_width", 2048),
                    "max_height": info.get("max_height", 2048),
                    "min_width": 256,
                    "min_height": 256,
                    "supported_formats": ["jpg", "jpeg", "png"],
                    "supports_img2img": info.get("supports_img2img", False),
                    "supports_seed": True,
                    "supports_enhancement": True
                }
            else:
                # Structured model data
                normalized_model = {
                    "name": model.get("name", "unknown"),
                    "description": model.get("description", ""),
                    "style_tags": model.get("style_tags", []),
                    "max_width": model.get("max_width", 2048),
                    "max_height": model.get("max_height", 2048),
                    "min_width": model.get("min_width", 256),
                    "min_height": model.get("min_height", 256),
                    "supported_formats": model.get("supported_formats", ["jpg", "png"]),
                    "supports_img2img": model.get("supports_img2img", False),
                    "supports_seed": model.get("supports_seed", True),
                    "supports_enhancement": model.get("supports_enhancement", True)
                }

            normalized.append(normalized_model)

        return normalized


def main():
    """Example usage of model retrieval"""
    print("=" * 60)
    print("Model Retrieval Examples")
    print("=" * 60)

    retriever = ModelRetrieval()

    # List text models
    print("\n1. Listing Text Models:")
    print("-" * 60)
    text_models = retriever.list_text_models()
    for model in text_models:
        print(f"\nModel: {model['name']}")
        print(f"  Description: {model['description']}")
        print(f"  Max Tokens: {model['max_input_tokens']}")
        print(f"  Reasoning: {model['reasoning_capable']}")
        print(f"  Vision: {model['vision']}")
        print(f"  Audio: {model['audio']}")
        print(f"  Tool Use: {model['tool_use']}")
        if model['voices']:
            print(f"  Voices: {', '.join(model['voices'])}")

    # List image models
    print("\n\n2. Listing Image Models:")
    print("-" * 60)
    image_models = retriever.list_image_models()
    for model in image_models:
        print(f"\nModel: {model['name']}")
        print(f"  Description: {model['description']}")
        print(f"  Style Tags: {', '.join(model['style_tags'])}")
        print(f"  Max Size: {model['max_width']}x{model['max_height']}")
        print(f"  Formats: {', '.join(model['supported_formats'])}")
        print(f"  Image-to-Image: {model['supports_img2img']}")

    # Export to JSON
    print("\n\n3. Exporting model data to JSON:")
    print("-" * 60)
    with open("text_models.json", "w") as f:
        json.dump(text_models, f, indent=2)
    print("Text models saved to text_models.json")

    with open("image_models.json", "w") as f:
        json.dump(image_models, f, indent=2)
    print("Image models saved to image_models.json")


if __name__ == "__main__":
    main()

File: PolliLibJS/model-retrieval.js
/**
 * Model Retrieval - List available text and image models
 * Implements the Model Retrieval section from the TODO list
 */

const { PollinationsAPI } = require('./pollylib');

/**
 * Class for retrieving available models from Pollinations.AI
 */
class ModelRetrieval extends PollinationsAPI {
    /**
     * List all available text generation models.
     *
     * @param {Object} options - Options
     * @param {boolean} options.normalized - Return normalized model schema with full details
     * @returns {Promise<Array>} List of model information dictionaries
     */
    async listTextModels(options = {}) {
        const { normalized = true } = options;

        try {
            const response = await this.retryRequest(
                `${PollinationsAPI.TEXT_API}/models`,
                { method: "GET" }
            );

            const models = await response.json();

            if (normalized) {
                return this._normalizeTextModels(models);
            }

            return models;

        } catch (error) {
            console.error(`Error retrieving text models: ${error.message}`);
            return [];
        }
    }

    /**
     * Normalize text model data into standard schema.
     *
     * @param {*} models - Raw model data
     * @returns {Array} Normalized model list
     */
    _normalizeTextModels(models) {
        const normalized = [];

        // Handle different response formats
        let modelList;
        if (Array.isArray(models)) {
            modelList = models;
        } else if (typeof models === 'object' && models.models) {
            modelList = models.models;
        } else {
            return [];
        }

        for (const model of modelList) {
            let normalizedModel;

            if (typeof model === 'string') {
                // Basic model name only
                normalizedModel = {
                    name: model,
                    description: `${model} text generation model`,
                    max_input_tokens: 128000,
                    reasoning_capable: model.toLowerCase().includes('reasoning'),
                    tier: 'standard',
                    community_supported: false,
                    input_types: ['text'],
                    output_types: ['text'],
                    tool_use: model.toLowerCase().includes('openai'),
                    aliases: [],
                    vision: model.toLowerCase().includes('vision') || ['openai', 'openai-large', 'claude-hybridspace'].includes(model),
                    audio: model.toLowerCase().includes('audio'),
                    voices: model.toLowerCase().includes('audio') ? ['alloy', 'echo', 'fable', 'onyx', 'nova', 'shimmer'] : [],
                    system_messages_supported: true,
                    uncensored: false
                };
            } else {
                // Structured model data
                normalizedModel = {
                    name: model.name || 'unknown',
                    description: model.description || '',
                    max_input_tokens: model.max_input_tokens || 128000,
                    reasoning_capable: model.reasoning_capable || false,
                    tier: model.tier || 'standard',
                    community_supported: model.community_supported || false,
                    input_types: model.input_types || ['text'],
                    output_types: model.output_types || ['text'],
                    tool_use: model.tool_use || false,
                    aliases: model.aliases || [],
                    vision: model.vision || false,
                    audio: model.audio || false,
                    voices: model.voices || [],
                    system_messages_supported: model.system_messages_supported !== undefined ? model.system_messages_supported : true,
                    uncensored: model.uncensored || false
                };
            }

            normalized.push(normalizedModel);
        }

        return normalized;
    }

    /**
     * List all available image generation models.
     *
     * @param {Object} options - Options
     * @param {boolean} options.normalized - Return normalized model schema with full details
     * @returns {Promise<Array>} List of model information dictionaries
     */
    async listImageModels(options = {}) {
        const { normalized = true } = options;

        try {
            const response = await this.retryRequest(
                `${PollinationsAPI.IMAGE_API}/models`,
                { method: "GET" }
            );

            const models = await response.json();

            if (normalized) {
                return this._normalizeImageModels(models);
            }

            return models;

        } catch (error) {
            console.error(`Error retrieving image models: ${error.message}`);
            return [];
        }
    }

    /**
     * Normalize image model data into standard schema.
     *
     * @param {*} models - Raw model data
     * @returns {Array} Normalized model list
     */
    _normalizeImageModels(models) {
        const normalized = [];

        // Handle different response formats
        let modelList;
        if (Array.isArray(models)) {
            modelList = models;
        } else if (typeof models === 'object' && models.models) {
            modelList = models.models;
        } else {
            return [];
        }

        // Known model characteristics
        const modelInfo = {
            flux: {
                description: 'High-quality image generation model',
                style_tags: ['photorealistic', 'artistic', 'detailed'],
                max_width: 2048,
                max_height: 2048,
                supports_img2img: false
            },
            turbo: {
                description: 'Fast image generation model',
                style_tags: ['quick', 'artistic'],
                max_width: 1024,
                max_height: 1024,
                supports_img2img: false
            },
            kontext: {
                description: 'Image-to-image transformation model',
                style_tags: ['transformation', 'editing'],
                max_width: 2048,
                max_height: 2048,
                supports_img2img: true
            }
        };

        for (const model of modelList) {
            let normalizedModel;

            if (typeof model === 'string') {
                const modelName = model;
                const info = modelInfo[modelName] || {};

                normalizedModel = {
                    name: modelName,
                    description: info.description || `${modelName} image model`,
                    style_tags: info.style_tags || ['general'],
                    max_width: info.max_width || 2048,
                    max_height: info.max_height || 2048,
                    min_width: 256,
                    min_height: 256,
                    supported_formats: ['jpg', 'jpeg', 'png'],
                    supports_img2img: info.supports_img2img || false,
                    supports_seed: true,
                    supports_enhancement: true
                };
            } else {
                // Structured model data
                normalizedModel = {
                    name: model.name || 'unknown',
                    description: model.description || '',
                    style_tags: model.style_tags || [],
                    max_width: model.max_width || 2048,
                    max_height: model.max_height || 2048,
                    min_width: model.min_width || 256,
                    min_height: model.min_height || 256,
                    supported_formats: model.supported_formats || ['jpg', 'png'],
                    supports_img2img: model.supports_img2img || false,
                    supports_seed: model.supports_seed !== undefined ? model.supports_seed : true,
                    supports_enhancement: model.supports_enhancement !== undefined ? model.supports_enhancement : true
                };
            }

            normalized.push(normalizedModel);
        }

        return normalized;
    }
}

// Example usage
async function main() {
    console.log("=".repeat(60));
    console.log("Model Retrieval Examples");
    console.log("=".repeat(60));

    const retriever = new ModelRetrieval();

    // List text models
    console.log("\n1. Listing Text Models:");
    console.log("-".repeat(60));
    const textModels = await retriever.listTextModels();
    for (const model of textModels) {
        console.log(`\nModel: ${model.name}`);
        console.log(`  Description: ${model.description}`);
        console.log(`  Max Tokens: ${model.max_input_tokens}`);
        console.log(`  Reasoning: ${model.reasoning_capable}`);
        console.log(`  Vision: ${model.vision}`);
        console.log(`  Audio: ${model.audio}`);
        console.log(`  Tool Use: ${model.tool_use}`);
        if (model.voices.length > 0) {
            console.log(`  Voices: ${model.voices.join(', ')}`);
        }
    }

    // List image models
    console.log("\n\n2. Listing Image Models:");
    console.log("-".repeat(60));
    const imageModels = await retriever.listImageModels();
    for (const model of imageModels) {
        console.log(`\nModel: ${model.name}`);
        console.log(`  Description: ${model.description}`);
        console.log(`  Style Tags: ${model.style_tags.join(', ')}`);
        console.log(`  Max Size: ${model.max_width}x${model.max_height}`);
        console.log(`  Formats: ${model.supported_formats.join(', ')}`);
        console.log(`  Image-to-Image: ${model.supports_img2img}`);
    }

    // Export to JSON
    console.log("\n\n3. Exporting model data to JSON:");
    console.log("-".repeat(60));
    const fs = require('fs').promises;
    await fs.writeFile('text_models.json', JSON.stringify(textModels, null, 2));
    console.log("Text models saved to text_models.json");

    await fs.writeFile('image_models.json', JSON.stringify(imageModels, null, 2));
    console.log("Image models saved to image_models.json");
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { ModelRetrieval };
}

// Test if run directly
if (typeof require !== 'undefined' && require.main === module) {
    main().catch(console.error);
}
